### PR TITLE
feat: Add form filling use case

### PR DESCRIPTION
**Pull request by leehuwuj**
## Summary by CodeRabbit

- **New Features**
	- Introduced a "Form Filling" tool to enhance data entry capabilities.
	- Added a new application type for "form_filling" in user prompts.
	- Implemented a workflow for programmatically filling out forms using various tools and events.
	- Improved event handling for progress updates in the chat interface.
	- Enhanced the `ChatAgentEvents` component to manage and display progress alongside text messages.
	- Added a customizable progress bar component for better user feedback.
	- Introduced new event classes to streamline the form-filling workflow.
	- Added functionality for agent selection based on user prompts.
	- Enhanced file uploader to support multiple file uploads.

- **Bug Fixes**
	- Updated handling of configured tools for document generation and report capabilities.

- **Documentation**
	- Added a README template for the form filling project, outlining setup and usage instructions.

- **Chores**
	- Updated dependencies to include a new progress bar library for enhanced UI features.

### CODE REVIEW

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts [hunk 1]
```diff
@@ -57,8 +57,9 @@ export type EventData = {
 };
 
 export type AgentEventData = {
-  agent: string;
-  text: string;
+  name: string;
+  msg: string;
+  event_type: "text" | "progress";
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/index.ts:62]: alternative format (TBD):
we're sending an event with an update of the status of a task (that sends its feedback as text)
text and that 
```suggestion
  agent: string;
  type: "task"; // can be extended with other types
  data: AgentTask
}
type AgentTask = {
  id: string; // (I guess you need this to group the progress)
  msg: string;
  progress?: ProgressData;
}
type ProgressData = {
  current: number;
  total: number;
}
```

File: templates/types/streaming/fastapi/app/services/file.py [hunk 2]
```diff
@@ -79,9 +79,16 @@ def process_private_file(
...
+        csv_tools = [
+            "interpreter",
+            "artifact",
+            "fill_form",
+            "extract_questions",
+            "form_filling",
+        ]
         # If the file is CSV and there is a code executor tool, we don't need to index.
-        if extension == "csv" and any(tool in tools for tool in code_executor_tools):
+        if extension == "csv" and any(tool in tools for tool in csv_tools):
```
[the remark this answers is not in the document]
Comment (leehuwuj) [reply templates/types/streaming/fastapi/app/services/file.py]: agree!

**Comment by leehuwuj** [comment]
@coderabbitai review

File: questions/simple.ts [hunk 3]
```diff
@@ -152,6 +157,14 @@ const convertAnswers = async (
       frontend: true,
       modelConfig: MODEL_GPT4o,
     },
+    form_filling: {
+      template: "multiagent",
+      agents: "form_filling",
+      tools: getTools(["form_filling"]),
+      dataSources: EXAMPLE_10K_SEC_FILES,
```
[the remark this answers is not in the document]
Comment (leehuwuj) [reply questions/simple.ts:164]: Yes, if we stick with pdf upload use case then the dataSource should be empty and we can add the link to the PDF files in the README

[the remark this answers is not in the document]
Comment (marcusschiesser) [reply questions/simple.ts:164]: better we use then only one PDF

File: templates/components/agents/python/form_filling/app/engine/engine.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from app.engine.index import get_index
+from app.engine.tools import ToolFactory
+from app.engine.workflow import FormFillingWorkflow
+from llama_index.core.chat_engine.types import ChatMessage
+from llama_index.core.indices.vector_store import VectorStoreIndex
+from llama_index.core.tools import QueryEngineTool
+from llama_index.core.workflow import Workflow
+
+
+def get_chat_engine(
```
[the remark this answers is not in the document]
Comment (leehuwuj) [reply templates/components/agents/python/form_filling/app/engine/engine.py:8]: I just let the code as other agent template and i will have another ticket to optimize the code for all agents later.

File: templates/components/multiagent/python/app/api/routers/vercel_response.py [hunk 5]
```diff
@@ -55,32 +54,26 @@ def _create_stream(
         self,
         request: Request,
         chat_data: ChatData,
-        event_handler: Awaitable[AgentRunResult | AsyncGenerator],
-        events: AsyncGenerator[AgentRunEvent, None],
+        event_handler: Awaitable,
```
Comment (marcusschiesser) [inline templates/components/multiagent/python/app/api/routers/vercel_response.py:57]: why remove the type hints?

File: templates/types/streaming/fastapi/app/services/file.py [hunk 6]
```diff
@@ -78,17 +77,8 @@ def process_private_file(
...
-        csv_tools = [
-            "interpreter",
-            "artifact",
-            "fill_form",
-            "extract_questions",
-            "form_filling",
-        ]
-        # If the file is CSV and there is a code executor tool, we don't need to index.
-        if extension == "csv" and any(tool in tools for tool in csv_tools):
+        # Don't index csv file if csv tools are available
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/services/file.py:80]:
```suggestion
        # Don't index csv files (they are handled by tools)
```

File: questions/simple.ts [hunk 7]
```diff
@@ -35,6 +36,10 @@ export const askSimpleQuestions = async (
           title: "Financial Report Generator (using Workflows)",
           value: "financial_report_agent",
         },
+        {
+          title: "Form filling (using Workflows)",
+          value: "form_filling",
+        },
         { title: "Code Artifact Agent", value: "code_artifact" },
         { title: "Structured extraction", value: "extractor" },
```
Comment (marcusschiesser) [inline questions/simple.ts:44]:
```suggestion
        { title: "Information Extractor", value: "extractor" },
```

File: questions/simple.ts [hunk 8]
```diff
@@ -35,6 +36,10 @@ export const askSimpleQuestions = async (
           title: "Financial Report Generator (using Workflows)",
           value: "financial_report_agent",
         },
+        {
+          title: "Form filling (using Workflows)",
```
Comment (marcusschiesser) [inline questions/simple.ts:40]:
```suggestion
          title: "Form Filler (using Workflows)",
```

File: templates/components/agents/python/form_filling/app/agents/form_filling.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+
+def create_workflow(
+    chat_history: Optional[List[ChatMessage]] = None, **kwargs
+) -> Workflow:
+    index: VectorStoreIndex = get_index()
+    if index is None:
+        raise ValueError(
+            "Index is not found! Please run `poetry run generate` to create an index."
+        )
```
Comment (marcusschiesser) [inline templates/components/agents/python/form_filling/app/agents/form_filling.py]: error: readme doesn't require to do so

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts [hunk 10]
```diff
@@ -57,8 +57,9 @@ export type EventData = {
 };
 
 export type AgentEventData = {
-  agent: string;
-  text: string;
+  name: string;
+  msg: string;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/index.ts:61]: if we change the attributes, we also need to update the TS backend, right?

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts (hunk 1 above)
Comment (marcusschiesser) [reply templates/types/streaming/nextjs/app/components/ui/chat/index.ts:62]: to keep changes minimal, how about we move `ProgressData` to here and make this change: 
```suggestion
  agent: string;
  data: string | ProgressData;
  type: "text" | "progress";
```

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 11]
```diff
@@ -20,10 +21,18 @@ const AgentIcons: Record<string, LucideIcon> = {
...
+  total_steps: number;
+  current_step: number;
+  step_msg: string;
+};
+
 type MergedEvent = {
   agent: string;
   texts: string[];
   icon: LucideIcon;
+  progress?: ProgressData;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:35]: isn't that the right type?
```suggestion
  icon: LucideIcon;
  data: Array<string | ProgressData>;
```

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 12]
```diff
@@ -20,10 +21,18 @@ const AgentIcons: Record<string, LucideIcon> = {
   publisher: icons.BookCheck,
 };
 
+type ProgressData = {
+  progress_id: string;
+  total_steps: number;
+  current_step: number;
+  step_msg: string;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: snake_case is weird in TS, how about avoiding it?
```suggestion
  id: string;
  total: number;
  current: number;
  msg: string;
```

File: templates/components/agents/python/form_filling/app/agents/form_filling.py [hunk 13]
```diff
@@ -213,6 +212,16 @@ async def extract_missing_cells(
         )
         self.memory.put(message)
 
+        if self.query_engine_tool is None:
+            # Fallback to input that query engine tool is not found so that cannot answer questions
+            self.memory.put(
+                ChatMessage(
+                    role=MessageRole.USER,
+                    content="Extracted missing cells but query engine tool is not found so cannot answer questions. Ask user to upload file or connect to a knowledge base.",
```
Comment (marcusschiesser) [inline templates/components/agents/python/form_filling/app/agents/form_filling.py:220]: very good!

File: templates/components/multiagent/python/app/api/routers/vercel_response.py (hunk 5 above)
Comment (leehuwuj) [reply templates/components/multiagent/python/app/api/routers/vercel_response.py:57]: we have a type checking for the result of the event handler already. I removed the hint to not import anything from single.py code in the vercel response.

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts (hunk 10 above)
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/index.ts]: Great catch! We should keep this attributes as is to keep the compatible with TS backend.

Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/index.ts]: Here is now the typing:

```typescript
export type ProgressData = {
  id: string;
  total: number;
  current: number;
};

export type AgentEventData = {
  agent: string;
  text: string;
  type: "text" | "progress";
  data?: ProgressData;
};

```

I moved the `ProgressData` to index.ts , remove the snake case pattern and reuse the old attribute names to keep compatible with TS backend. It's minimal and enough for this PR and we can enhance it when implementing for TS

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts (hunk 1 above)
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/index.ts]: Update, please check my comment: https://github.com/run-llama/create-llama/pull/403#discussion_r1821984741

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx (hunk 11 above)
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: Yes, the type is correct. I just add a new progress field if it has progress bar.

Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: My idea was it's still the normal text event and if the event has progress properties then we will render the progress bar, i think easier to handle with that way and don't need to check what is the type of the data, additionally, we still have the whole text events of the agent after all and user can review what are the events of the progress.

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 14]
```diff
@@ -61,8 +122,18 @@ function AgentEventContent({
...
 }) {
-  const { agent, texts } = event;
+  const { agent, texts, progress } = event;
   const AgentIcon = event.icon;
+  const [progressValue, setProgressValue] = useState(0);
+
+  useEffect(() => {
+    if (progress) {
+      // Add 1 to current_step to match the display value
+      const value = Math.round(((progress.current + 1) / progress.total) * 100);
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:132]: Can't we just calculate that in ProgressContent?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 15]
```diff
@@ -61,8 +122,18 @@ function AgentEventContent({
   isLast: boolean;
   isFinished: boolean;
 }) {
-  const { agent, texts } = event;
+  const { agent, texts, progress } = event;
   const AgentIcon = event.icon;
+  const [progressValue, setProgressValue] = useState(0);
+
+  useEffect(() => {
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:129]: Why do we need use effect? Seems like value is purely alculated

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx (hunk 11 above)
Comment (marcusschiesser) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: So progress is always implicitly just for the last message?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 16]
```diff
@@ -136,15 +200,41 @@ function mergeAdjacentEvents(events: AgentEventData[]): MergedEvent[] {
...
+        } else {
+          mergedEvents.push({
+            agent: event.agent,
+            texts: [],
+            icon: AgentIcons[event.agent.toLowerCase()] ?? icons.Bot,
+            progress: progressData,
+          });
+        }
+      } catch (e) {
+        console.error(
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:220]: When does this happen?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx [hunk 17]
```diff
@@ -136,15 +200,41 @@ function mergeAdjacentEvents(events: AgentEventData[]): MergedEvent[] {
...
+          event.text,
+        );
+      }
+      continue;
+    }
+
+    // Handle regular text events
     if (lastMergedEvent && lastMergedEvent.agent === event.agent) {
-      // If the last event in mergedEvents has the same non-null agent, add the title to it
       lastMergedEvent.texts.push(event.text);
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:232]: Seems way easier to me to just check for progress here and if it exists adding it

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx (hunk 16 above)
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx]: Great catch, removed!

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx (hunk 11 above)
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: I think you're talking about this one: https://github.com/run-llama/create-llama/pull/403#discussion_r1822043308
If it's then it's updated.

Comment (marcusschiesser) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-agent-events.tsx:28]: @leehuwuj it looks way better now, we can close this

File: templates/components/agents/python/financial_report/app/agents/analyst.py [hunk 18]
```diff
@@ -22,8 +22,8 @@ def _get_analyst_params() -> Tuple[List[type[FunctionTool]], str, str]:
     description = "Expert in analyzing financial data"
     configured_tools = ToolFactory.from_env(map_result=True)
     # Check if the interpreter tool is configured
-    if "interpreter" in configured_tools.keys():
-        tools.extend(configured_tools["interpreter"])
+    if "interpret" in configured_tools.keys():
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/components/agents/python/financial_report/app/agents/analyst.py:25]: (i didn't find "interpret" anywhere)

File: templates/types/streaming/nextjs/app/components/ui/file-uploader.tsx [hunk 19]
```diff
@@ -50,30 +52,51 @@ export default function FileUploader({
...
-    if (isFileSizeExceeded(file)) {
-      return onFileUploadError(
-        `File size exceeded. Limit is ${fileSizeLimit / 1024 / 1024} MB`,
-      );
+    for (const file of files) {
+      const fileExtension = file.name.split(".").pop() || "";
+      const extensionFileError = checkExtension(fileExtension);
+      if (extensionFileError) {
+        onFileUploadError(extensionFileError);
+        continue;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/file-uploader.tsx:83]: Do we need to process more files if there's an error?

File: templates/types/streaming/nextjs/app/components/ui/file-uploader.tsx [hunk 20]
```diff
@@ -50,30 +52,51 @@ export default function FileUploader({
...
+      }
+
+      if (isFileSizeExceeded(file)) {
+        onFileUploadError(
+          `File size exceeded. Limit is ${fileSizeLimit / 1024 / 1024} MB`,
+        );
+        continue;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/file-uploader.tsx:90]: Do we need to process more files if there's an error?

Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/file-uploader.tsx]: Updated. Great catch! I thought i committed from coderabbit's comment already.

**Pull request merged**